content/developers/developer-patterns/navigating-merklelogs/index.md

```diff
@@ -0,0 +1,745 @@
+| 1   |     8      |          |      2   |    4   |      1      |     4    | count of bytes
+```
+
+The idtimestamp of the last leaf entry added to the log is always set in the header field.
+
+In the hex data above, the idtimestamp of the last entry in the log is `8e84dbbb6513a6`[^2], the version is `0`, the timestamp epoch is `1`, the massif height is `14`, and the massif index is `0`.
+
+[^2]: The idtimetamp value is 64 bits, of which the first 40 bits are a millisecond precision time value and the remainder is data used to guarantee uniqueness of the timestamp.
+DataTrails uses a variant of the [Snowflake ID](https://en.wikipedia.org/wiki/Snowflake_ID) scheme.
+The DataTrails implementation can be found at [nextid.go](https://github.com/datatrails/go-datatrails-merklelog/blob/main/massifs/snowflakeid/nextid.go#L118)
+
+### Decoding an idtimestamp
+
+The idtimestamp is 40 bits of time at millisecond precision.
+The idtimestamp in the header field is always set to the idtimestamp of the most recently added leaf.
+
+{{< tabs >}}
+   {{< tab name="Python" >}}
+
+   ```python
+   import datetime
+
+   epoch=1
+   unixms=int((
+      bytes.fromhex("8e84dbbb6513a600")[:-3]).hex(), base=16
+      ) + epoch*((2**40)-1)
+   datetime.datetime.fromtimestamp(unixms/1000)
+   ```
+
+   Generates:
+
+   ```output
+   datetime.datetime(2024, 3, 28, 11, 39, 36, 676000)
+   ```
+
+  {{< /tab >}}
+{{< /tabs >}}
+
+In this example, the last entry in the log (at that time) was `2024/03/28`, a little after 11.30am.
+
+## The trieData entries are 512 bits each and are formed from two fields
+
+With the trie keys, full data recovery can be verified without needing to remember the order that events were added to the log.
+
+Massifs can be recovered independently even if the data for other massifs is incomplete.
+
+The index also provides for proof of exclusion. Proof that a specific event is not in the index. As the event identities contribute to both the index keys and the leaf entries, and as the keys are time ordered on addition, it is possible to audit the specific range.
+This will confirm the log is consistent with the index for the time range in which the disputed item is claimed to exist.
+
+The trieData section is 2 \* 32 \* 2<sup>height</sup> bytes long, which is exactly double what is needed.
+For a standard massif height of 14, it has 8,192 entries in the first 524,288 bytes.
+The subsequent 524,288 will always be zero.
+The format of each entry is then, for a massif height of 14:
+
+```output
++----------------+
+| HEADER DATA    |
++----------------+
+
+288
++----------------+
+| TRIE-DATA      |
+|                |
+|                |
+|................| 288 + 524288 = 524576
+|  always        |
+|                |
+|       zero     |
++----------------+ 288 + 524288 * 2 = 1048864
+```
+
+Each entry is formatted like this
+
+```output
+0                                                        31
+SHA256(BYTE(0x00) || tenant-identity || event.identity)
```

Comment:
@robinbryce was this the formatting you were looking for?
Was this supposed to be a table looking visual?